src/multi.c

```diff
@@ -352,13 +358,30 @@ void touchWatchedKey(redisDb *db, robj *key) {
     /* Check if we are already watching for this key */
     listRewind(clients,&li);
     while((ln = listNext(&li))) {
-        client *c = listNodeValue(ln);
+        watchedKey *wk = listNodeValue(ln);
+        client *c = wk->client;
+
+        if (wk->expired) {
+            /* The key was already expired when WATCH was called. */
+            if (db == wk->db &&
+                equalStringObjects(key, wk->key) &&
+                dictFind(db->dict, key->ptr) == NULL)
+            {
+                /* Already expired key is deleted, so logically no change. */
+                wk->expired = 0;
+                goto skip_client;
```

Comment:
Yes, but I don't think we need to clear it actually. If the key doesn't exist, it is touched only when it's created.

Reply:
i'd rather be more explicit, and normalize the state so we don't have surprises some day in the future.
so when we're watching a missing key, we want that flag off.
it could make the code in touchAllWatchedKeysInDb more complicated, but also more explicit.